jane/doc/extensions/_04-modes/syntax.md

```diff
@@ -100,40 +115,126 @@ You can also specify the mode of the function body:
+`contended` modality applied to an `uncontended` record yields a `contended`
+field, because `uncontended < contended`.
+
+Any axis left out of a modality expression is assumed to be the identity
+modality. (When a modality is the identity, then the mode of a field is the same
+as the mode of the record.) For future axes, this would be the top mode; for
+past axes, this would be the bottom mode. These are the identity modalities:
+
+```ocaml
+local unique once nonportable uncontended unyielding stateless immutable
+```
+
+Note that a legacy mode might or might not be the same as the identity modality.
+
+It is an error to specify more than one modality along the same axis in one modality
+expression.
+
+All modality expressions are prefixed by an `@@` symbol,
+in one of several places in the syntax, as described below.
 
 ## Record fields
-Record fields can have modalities, for example:
+Record fields can have modalities:
 ```ocaml
 type r = {x : string @@ modalities}
 type r = {x : string @ modes -> string @ modes @@ modalities}
 ```
 
+## Constructor fields
+Constructor fields can have modalities:
+```ocaml
+type v =
+  | K1 of string @@ modalities
+  | K2 of string @@ modalities * int array
+  | K3 of string @@ modalities * int array @@ modalities
+  | K4 of (int -> int) @@ modalities   (* parentheses around functions are required even without modalities *)
+  | K5 : string @@ modalities -> v
+  | K6 : string @@ modalities * int array @@ modalities -> v
+  | K7 of { x : string @@ modalities; y : string @@ modalities }
+  | K8 : { x : string @@ modalities; y : string @@ modalities } -> v
+```
+
 ## Signature items
-Signature items such as values can have modalities; for example:
+A `val` signature item can have modalities:
 ```ocaml
 val foo : string @@ modalities
 val bar : string @ modes -> string @ modes @@ modalities
 ```
 
-A signature can have default modalities that each item can override; for example:
+Similarly, so can an `external` signature item:
+```ocaml
+external id : 'a -> 'a @@ modalities = "%identity"
+```
+
+A signature can have default modalities that each item can override:
 ```ocaml
 sig @@ portable
 val foo : string (* have portable modality *)
 val bar : string -> string @@ nonportable (* not have portable modality *)
 end
 ```
+These default modalities must be the first item in the signature.
+
+An .mli file is like a signature, but we do not write the `sig` and the
+`end`. Accordingly, you may put `@@ modalities` as the first item in an .mli
+file.
+
+<!-- CR reisenberg for zqian: There are other signature items that can have
+modalities, according to the parser:
+  include S @@ modalities
+  module (M @@ modalities) : module_type   (* but this does not work with [rec] *)
+  module M : module_type @@ modalities  (* this form *does* work with [rec] *)
+  module (M @@ modalities) = M2
+  module M = M2 @@ modalities
+I don't exactly know what these mean, so I have not documented them.
+-->
```

Comment:
The first is kinda hacky right now, and I'd rather not mention it until #3759 .

The rest are modules which are not supported yet.